### ISSUE TITLE
CRASH in notepad on Win10-1903

### ISSUE THREAD

**Issue by Gurtern** [DynamoRIO/dynamorio#4289]
steps to reproduce-
1. build master (debug and release), or obtain build. Have tried 'DynamoRIO-Windows-8.0.18390' , 
2. bin32\drrun.exe -debug   -- notepad

3. result
![drcov-notepad](https://user-images.githubusercontent.com/19630618/81610674-c2bc9780-938e-11ea-9ff7-99319cca17e1.png)

![winver](https://user-images.githubusercontent.com/19630618/81610950-22b33e00-938f-11ea-8534-623c8c3c3d82.png)

**Comment by derekbruening** [comment]
I cannot reproduce, and it is likely that this only happens on particular environments, or else it would have been observed much earlier.

I would suggest some debugging:
+ Which set of applications and bitwidths does this happen on: does it happen on hello,world?  32-bit on 64-bit-kernel only?
+ If you have access to other Win10 machines with different software installed, does it reproduce there?
+ Do you have invasive security software on this machine injecting libraries into every process?  Look at the library list in notepad.exe at the point of this popup window: what is in there that is not a standard Windows system library?
+ Attach a debugger like windbg and examine the crash point to try to see what is going on.

**Comment by Gurtern** [comment]
> Which set of applications and bitwidths does this happen on: does it happen on hello,world? 32-bit on 64-bit-kernel only?

32-bit. 

> If you have access to other Win10 machines with different software installed, does it reproduce there?
Reproduced on build 'DynamoRIO-Windows-8.0.18390' in 'MS Edge Win10 preview' virtual machine (Virtual Box). It appears to be erratic. I had about 2/10 successes for 'bin32\drrun.exe -32 -t drcov notepad'. Successes usually followed after attempting 'bin64\drrun.exe -32 -t drcov notepad', which did not work. Which begs the greater question- should I be using 'bin64\drrun.exe -32' on win10 to run a 32-bit application. 

> Do you have invasive security software
 Only Windows Defender. Disabled Real-time Protection did not make a difference. 

> Attach a debugger
Attached debugger to drrun.exe 'bin32\drrun.exe -32 -t drcov -- notepad'. 
` 	[External Code]	
 	[Frames below may be incorrect and/or missing, no symbols loaded for ntdll.dll]	Unknown
>	drinjectlib.dll!nt_wait_event_with_timeout(void * hevent, _LARGE_INTEGER * timeout) Line 3752	C
 	drinjectlib.dll!os_wait_handle(void * h, unsigned __int64 timeout_ms) Line 8548	C
 	drinjectlib.dll!dr_inject_wait_for_child(void * data, unsigned __int64 timeout_millis) Line 951	C
 	drrun.exe!wmain(int argc, wchar_t * * targv) Line 1769	C
`
The crash is in notepad.exe. The Visual Studio debugger attaches to notepad.exe , but the debugger controls do not work.

**Comment by derekbruening** [comment]
A callstack of the drrun launcher is not likely to help: you want 
a callstack of the crashing child process.  I would repeat the earlier suggestion as well: minimize the app and run hello,world instead of notepad.
https://github.com/DynamoRIO/dynamorio/wiki/Debugging has further information.

**Comment by Gurtern** [comment]
>  .\bin32\drrun.exe -msgbox_mask 15 -t drcov -dump_text -- notepad

![access_violation_1](https://user-images.githubusercontent.com/19630618/82081606-737eab80-969b-11ea-85bc-3f5601fa57cc.png)
![access_violation_2](https://user-images.githubusercontent.com/19630618/82081613-78435f80-969b-11ea-9a31-107d55eaabba.png)

There is also a way forward, which involves inspecting the threads in WinDbg. Just selecting them or arrow-keying through the list seems to be adequate. Then if I were to hit f5 to continue the program, it oddly works.

**Comment by debach** [comment]
Some observations in case somebody is interested. If time permits, I will try to debug this.

I observe many crashes without any client as well. For example, `drrun.exe -debug -- notepad.exe` crashes on a debug build of e1028223 (built with VS 2017 x64 on Windows 10 x64 1903). The crash occurs when opening the *File Open* dialogue – which triggers *a lot* of module loads – in notepad, in `dispatch.c(757):dispatch_enter_dynamorio`:

    ASSERT(wherewasi == DR_WHERE_FCACHE || wherewasi == DR_WHERE_TRAMPOLINE ||
           wherewasi == DR_WHERE_APP ||
           /* If the thread was waiting at check_wait_at_safe_point when getting
            * suspended, we were in dispatch (ref i#3427). We will be here after the
            * thread's context is being reset before sending it native.
            */
           (dcontext->go_native && wherewasi == DR_WHERE_DISPATCH));

A non-debug build of e1028223 crashes only after I load a file in notepad.exe – again, many module loads and unloads are caused by this – in `native_exec.c(148):check_and_mark_native_exec`:

    static bool
    check_and_mark_native_exec(module_area_t *ma, bool add)
    {
        bool is_native = false;
        const char *name = GET_MODULE_NAME(&ma->names);

Here, `ma` is a null pointer, causing:

    …
    ModLoad: 00007ff9`796b0000 00007ff9`796ba000   C:\Windows\system32\FLTLIB.DLL
    shell\comdlg32\fileopensave.cpp(14403)\comdlg32.dll!00007FF97FE8A098: (caller: 00007FF97FEB8131) ReturnHr(1) tid(29b8) 80004005 Unbekannter Fehler
        CallContext:[\PickerModalLoop] 
    (1ce4.29b8): Access violation - code c0000005 (first chance)
    First chance exceptions are reported before any exception handling.
    This exception may be expected and handled.
    dynamorio!check_and_mark_native_exec+0x12:
    00000000`71048cf2 488b5920        mov     rbx,qword ptr [rcx+20h] ds:00000000`00000020=????????????????

Stack:

    1:004> k
     # Child-SP          RetAddr           Call Site
    00 00000252`f206dfc0 00000000`71049139 dynamorio!check_and_mark_native_exec+0x12 [c:\projects\dynamorio\core\native_exec.c @ 148] 
    01 00000252`f206dff0 00000000`7103cad0 dynamorio!native_exec_module_unload+0x19 [c:\projects\dynamorio\core\native_exec.c @ 189] 
    02 00000252`f206e020 00000000`710a7d3d dynamorio!module_list_remove+0xc0 [c:\projects\dynamorio\core\module_list.c @ 336] 
    03 (Inline Function) --------`-------- dynamorio!process_image_post_vmarea+0x1c [c:\projects\dynamorio\core\win32\os.c @ 4332] 
    04 00000252`f206e050 00000000`710ad335 dynamorio!process_mmap+0x1dd [c:\projects\dynamorio\core\win32\os.c @ 4520] 
    05 (Inline Function) --------`-------- dynamorio!presys_UnmapViewOfSection+0xf0 [c:\projects\dynamorio\core\win32\syscall.c @ 2701] 
    06 00000252`f206e0f0 00000000`7101ede3 dynamorio!pre_system_call+0xd45 [c:\projects\dynamorio\core\win32\syscall.c @ 2984] 
    07 00000252`f206eec0 00000000`7101d884 dynamorio!handle_system_call+0x1f3 [c:\projects\dynamorio\core\dispatch.c @ 1988] 
    08 00000252`f206ef00 00000000`7101cf37 dynamorio!dispatch_enter_dynamorio+0x504 [c:\projects\dynamorio\core\dispatch.c @ 909] 
    09 00000252`f206ef30 00007ff6`af7625a6 dynamorio!d_r_dispatch+0x27 [c:\projects\dynamorio\core\dispatch.c @ 164] 
    0a 00000252`f206efe0 00000000`00000246 0x00007ff6`af7625a6
    0b 00000252`f206efe8 00000000`00000000 0x246

The assertions in `module_list_remove` are not triggered in the release version, but might be helpful:

    ma = (module_area_t *)vmvector_lookup(loaded_module_areas, base);
    ASSERT_CURIOSITY(ma != NULL); /* loader can't have a race */

The module in question is `MsftEdit.dll`.

At that time, there are 18 additional threads. Three threads are, presumably, for the parallel loading of DLLs (Windows 10 does that), and all have a shallow 4-entry call stack, the top being `dynamorio!interception_code_array+0x1432`. The other 15 additional threads have very different call stacks that look quite native (no DynamoRIO stuff showing), except that the top entry is always `0x00007ff6af762cc2`, which looks like code cache to me, but I’m not sure.

Right now I can only say that the same crash happens in d9fdabee and everything works fine in 0a3c7f40.

**Comment by byemypast** [comment]
I encounter exactly the same problem. 
environmnet:
```
Windows version 10.0.18363.959
DynamoRIO-Windows-8.0.0-1
x64 kernel with 32bits mode (notepad.exe runs on syswow64 simulation mode)
```

here's what windbg logged.
```
ModLoad: 74c80000 74ca5000   C:\WINDOWS\SysWOW64\IMM32.DLL
ModLoad: 78b60000 78cc0000   D:\crack\Tools\0day\DynamoRIO-Windows-8.0.0-1\lib32\release\dynamorio.dll
(4fe8.56a4): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
dynamorio!switch_modes_and_call+0x4e:
78c0f94b 5b              pop     ebx
1:004:x86> k
 # ChildEBP RetAddr  
00 0267ead8 78be7949 dynamorio!switch_modes_and_call+0x4e [C:\projects\dynamorio\build\build_release-32\core\CMakeFiles\drlibc.dir\drlibc\drlibc_x86.asm.obj.s @ 1962] 
01 0267eb08 78be77d7 dynamorio!os_take_over_wow64_extra+0x49 [c:\projects\dynamorio\core\win32\os.c @ 2433] 
02 0267f654 78be74ef dynamorio!os_take_over_thread+0x197 [c:\projects\dynamorio\core\win32\os.c @ 2575] 
03 0267f6a4 78b65776 dynamorio!os_take_over_all_unknown_threads+0x2cf [c:\projects\dynamorio\core\win32\os.c @ 2707] 
04 0267f6c8 78be044e dynamorio!dynamorio_take_over_threads+0xe6 [c:\projects\dynamorio\core\dynamo.c @ 2929] 
05 0267f6e0 78be0718 dynamorio!auto_setup+0x3e [c:\projects\dynamorio\core\arch\x86_code.c @ 198] 
06 00000000 00000000 dynamorio!dynamo_auto_start+0x8 [C:\projects\dynamorio\build\build_release-32\core\CMakeFiles\dynamorio.dir\arch\x86\x86.asm.obj.s @ 1600] 
1:004:x86> u eip
dynamorio!switch_modes_and_call+0x4e [C:\projects\dynamorio\build\build_release-32\core\CMakeFiles\drlibc.dir\drlibc\drlibc_x86.asm.obj.s @ 1962]:
78c0f94b 5b              pop     ebx
78c0f94c c3              ret
dynamorio!load_dynamo [C:\projects\dynamorio\build\build_release-32\core\CMakeFiles\drlibc.dir\drlibc\drlibc_x86.asm.obj.s @ 1981]:
78c0f94d 8bff            mov     edi,edi
78c0f94f 8944242c        mov     dword ptr [esp+2Ch],eax
78c0f953 5b              pop     ebx
78c0f954 58              pop     eax
78c0f955 83ec08          sub     esp,8
78c0f958 50              push    eax
1:004:x86> r
eax=00000000 ebx=0267ea08 ecx=6c57de04 edx=00000000 esi=18922e80 edi=000001b0
eip=78c0f94b esp=0267ea28 ebp=0267ead8 iopl=0         nv up ei pl nz na pe nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010206
dynamorio!switch_modes_and_call+0x4e:
78c0f94b 5b              pop     ebx

```

**Comment by derekbruening** [comment]
With the details of the crash showing that it's in switch_modes_and_call: this is a duplicate of #4091.  Let's move further discussion there.  Help is needed to fix the issue.  Closing as duplicate.
